### PR TITLE
Perform direct column replacements in `unnest_longer()` / `unnest_wider()`

### LINKED ISSUE #741: We can't use unnest_wider() on vctrs list columns
See #737 for full example

These are the columns that `pivot_wider()` creates so I don't believe this is expected : 

``` r
library(tidyr)

df1 <- tibble::tribble(
  ~type,          ~name, ~var1,
  "Country",   "Norway", 169L, 
    "Sport",     "Skii", 169L, 
  "Country",    "Spain", 150L, 
    "Sport",     "Bike", 150L, 
    "Sport",   "Soccer", 150L, 
    "Sport",   "Basket", 150L, 
  "Country",      "USA",   0L, 
    "Sport", "Baseball",   0L, 
)

df2 <- pivot_wider(df1, names_from = "type", values_from = "name") 
#> Warning: Values in `name` are not uniquely identified; output will contain list-cols.
#> * Use `values_fn = list(name = list)` to suppress this warning.
#> * Use `values_fn = list(name = length)` to identify where the duplicates arise
#> * Use `values_fn = list(name = summary_fun)` to summarise duplicates
df2
#> # A tibble: 3 x 3
#>    var1     Country       Sport
#>   <int> <list<chr>> <list<chr>>
#> 1   169         [1]         [1]
#> 2   150         [1]         [3]
#> 3     0         [1]         [1]

unnest_wider(df2, Sport)
#> New names:
#> * `` -> ...1
#> New names:
#> * `` -> ...1
#> * `` -> ...2
#> * `` -> ...3
#> New names:
#> * `` -> ...1
#> No common type for `x` <tbl_df<...1:character>> and `y` <character>.

df2$Sport <- unclass(df2$Sport)

unnest_wider(df2, Sport)
#> New names:
#> * `` -> ...1
#> New names:
#> * `` -> ...1
#> * `` -> ...2
#> * `` -> ...3
#> New names:
#> * `` -> ...1
#> # A tibble: 3 x 5
#>    var1     Country ...1     ...2   ...3  
#>   <int> <list<chr>> <chr>    <chr>  <chr> 
#> 1   169         [1] Skii     <NA>   <NA>  
#> 2   150         [1] Bike     Soccer Basket
#> 3     0         [1] Baseball <NA>   <NA>

pivot_wider(df1, names_from = "type", values_from = "name", values_fn = list(name=list))
#> # A tibble: 3 x 3
#>    var1     Country       Sport
#>   <int> <list<chr>> <list<chr>>
#> 1   169         [1]         [1]
#> 2   150         [1]         [3]
#> 3     0         [1]         [1]
```

<sup>Created on 2019-09-15 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### PR DESCRIPTION
**Pull request by DavisVaughan**
Fixes #741.

This was a similar problem to a previous issue, where we were assigning to `data[[col]][] <-` when it just needed to be `data[[col]] <- `.

I also changed this for `unnest_longer()`.

Note that there is still another place in `unnest_longer/wider()` where the `data[[col]][] <- ` assignment is used. That one was not adjusted, as it broke tests. I think it makes sense there because at that point `data[[col]]` is often a tibble itself (it was a df-col), and we want to use `data[[col]][] <-` there to perform the assignment, rather than replace the tibble.

``` r
library(tidyr)

df1 <- tibble::tribble(
  ~type,          ~name, ~var1,
  "Country",   "Norway", 169L, 
  "Sport",     "Skii", 169L, 
  "Country",    "Spain", 150L, 
  "Sport",     "Bike", 150L, 
  "Sport",   "Soccer", 150L, 
  "Sport",   "Basket", 150L, 
  "Country",      "USA",   0L, 
  "Sport", "Baseball",   0L, 
)

df2 <- pivot_wider(df1, names_from = "type", values_from = "name") 
#> Warning: Values in `name` are not uniquely identified; output will contain list-cols.
#> * Use `values_fn = list(name = list)` to suppress this warning.
#> * Use `values_fn = list(name = length)` to identify where the duplicates arise
#> * Use `values_fn = list(name = summary_fun)` to summarise duplicates
```

```r
# before
unnest_wider(df2, Sport)
#> New names:
#> * `` -> ...1
#> New names:
#> * `` -> ...1
#> * `` -> ...2
#> * `` -> ...3
#> New names:
#> * `` -> ...1
#> No common type for `x` <tbl_df<...1:character>> and `y` <character>.

# after
unnest_wider(df2, Sport)
#> New names:
#> * `` -> ...1
#> New names:
#> * `` -> ...1
#> * `` -> ...2
#> * `` -> ...3
#> New names:
#> * `` -> ...1
#> # A tibble: 3 x 5
#>    var1     Country ...1     ...2   ...3  
#>   <int> <list<chr>> <chr>    <chr>  <chr> 
#> 1   169         [1] Skii     <NA>   <NA>  
#> 2   150         [1] Bike     Soccer Basket
#> 3     0         [1] Baseball <NA>   <NA>
```

```r
# before
unnest_longer(df2, Sport)
#> No common type for `x` <tbl_df<Sport:character>> and `y` <character>.

# after
unnest_longer(df2, Sport)
#> # A tibble: 5 x 3
#>    var1     Country Sport   
#>   <int> <list<chr>> <chr>   
#> 1   169         [1] Skii    
#> 2   150         [1] Bike    
#> 3   150         [1] Soccer  
#> 4   150         [1] Basket  
#> 5     0         [1] Baseball
```

<sup>Created on 2019-09-16 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

### CODE REVIEW

File: tests/testthat/test-rectangle.R [hunk 1]
```diff
@@ -147,6 +147,14 @@ test_that("list of 0-length vectors yields no new columns", {
   expect_named(unnest_wider(df, y), "x")
 })
 
+test_that("list_of columns can be unnested", {
+  df <- tibble(x = 1:2, y = list_of(1L, 1:2))
```
Comment (hadley) [inline tests/testthat/test-rectangle.R]: Maybe make this `list_of(c(a = 1L), c(a = 1L, b = 2L))` to get better names in the output?

**Comment by hadley** [comment]
Oh and needs a news bullet too please

**Comment by DavisVaughan** [comment]
It felt important to mention the use cases where this arises, but I can simplify it back to just the first sentence of the bullet if it doesn't seem relevant

File: NEWS.md [hunk 2]
```diff
@@ -1,5 +1,12 @@
 # tidyr (development version)
 
+## Bug fixes and minor improvements
```
Comment (hadley) [inline NEWS.md]: Can you please remove this? We only add just prior to release (https://style.tidyverse.org/news.html#news-release)

**Pull request merged**